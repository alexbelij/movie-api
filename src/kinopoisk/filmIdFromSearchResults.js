/* @flow */

import cheerio from 'cheerio';
import R from 'ramda';

export type SearchQuery = {
  title: string,
  year?: number,
  countries?: Array<string>,
  isTvShow?: boolean,
};

type SearchResult = {
  id: number,
  title: string,
  countries: Array<string>,
  year: number,
};

const scrapeResults = (html: string): Array<SearchResult> => {
  const $ = cheerio.load(html);

  return $('.film-snippet')
    .get()
    .map((snippet: any) => {
      const titleNode = $(snippet).find('.film-snippet__title-link');
      const id = parseInt(
        R.nth(1, titleNode.attr('href').match(/\/film\/(.+)\//) || []),
        10,
      );

      if (isNaN(id)) return null;

      const infoNode = $(snippet).find('.film-snippet__info');
      const title = titleNode.text().replace(/\s\s+/g, ' ');
      const countries = R.pipe(
        R.split(','),
        R.map(R.trim),
        R.filter(
          (country: ?string) => !!country && isNaN(parseInt(country, 10)),
        ),
      )(infoNode.text());
      const year = parseInt(R.head(infoNode.text().match(/\d+/) || []), 10);

      return { id, title, countries, year };
    });
};

const filterResults = (
  query: SearchQuery,
  results: Array<?SearchResult>,
): Array<SearchResult> => R.filter(R.allPass([
  Boolean,
  (query.year
    ? ({ year }: SearchResult) => year === query.year
    : R.always(true)),
  (query.countries
    ? ({ countries }: SearchResult) => R.equals(
        R.intersection(countries, query.countries).length,
        // $FlowFixMe: countries is always available here
        query.countries.length,
      )
    : R.always(true)),
]))(results);

const filmIdFromSearchResults = (html: string, query: SearchQuery) => R.pipe(
  scrapeResults,
  R.curry(filterResults)(query),
  R.path(['0', 'id']),
)(html);

export {
  scrapeResults as __scrapeResults,
  filterResults as __filterResults,
};
export default filmIdFromSearchResults;
