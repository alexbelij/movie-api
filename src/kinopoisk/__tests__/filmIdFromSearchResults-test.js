/* @flow */

import { __filterResults } from '../filmIdFromSearchResults';

describe('kinopoisk/filmIdFromSearchResults', () => {
  it('filters results', () => {
    /* eslint-disable max-len */
    expect(__filterResults({ title: 'foo' }, [
      null,
      undefined,
      { id: 1, title: 'foo', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'Звёздные войны' }, [
      { id: 1, title: 'Звёздные войны', countries: [], year: 2015 },
      { id: 2, title: 'Звездные войны', countries: [], year: 2016 },
      { id: 3, title: 'Звёздная пыль', countries: [], year: 2014 },
      { id: 4, title: 'Звёздный путь', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({
      title: 'Star Wars: Episode VII - The Force Awakens',
    }, [
      { id: 1, title: 'Звёздные войны', originalTitle: 'Star Wars: Episode VI', countries: [], year: 2015 },
      { id: 2, title: 'Звёздные войны', originalTitle: 'Star Wars: Episode VII - The Force Awakens', countries: [], year: 2016 },
      { id: 3, title: 'Звёздные войны', originalTitle: 'Star Wars: Episode V', countries: [], year: 2014 },
      { id: 4, title: 'Звёздные войны', originalTitle: 'Star Wars: Episode IV', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'foo', year: 2016 }, [
      { id: 1, title: 'foo', countries: [], year: 2015 },
      { id: 2, title: 'foo', countries: [], year: 2016 },
      { id: 3, title: 'foo', countries: [], year: 2014 },
      { id: 4, title: 'foo', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'foo', countries: ['США', 'Россия'] }, [
      { id: 1, title: 'foo', countries: ['США'], year: 2016 },
      { id: 2, title: 'foo', countries: ['США', 'Россия', 'Германия'], year: 2016 },
      { id: 3, title: 'foo', countries: ['Россия'], year: 2016 },
      { id: 4, title: 'foo', countries: ['США', 'Россия'], year: 2016 },
    ])).toMatchSnapshot();
    /* eslint-enable max-len */
  });
});
