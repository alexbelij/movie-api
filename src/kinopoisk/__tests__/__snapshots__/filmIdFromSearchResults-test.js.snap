exports[`kinopoisk/filmIdFromSearchResults filters results 1`] = `
Array [
  Object {
    "countries": Array [],
    "id": 1,
    "title": "foo",
    "year": 2016,
  },
]
`;

exports[`kinopoisk/filmIdFromSearchResults filters results 2`] = `
Array [
  Object {
    "countries": Array [],
    "id": 1,
    "title": "Звёздные войны",
    "year": 2015,
  },
  Object {
    "countries": Array [],
    "id": 2,
    "title": "Звездные войны",
    "year": 2016,
  },
]
`;

exports[`kinopoisk/filmIdFromSearchResults filters results 3`] = `
Array [
  Object {
    "countries": Array [],
    "id": 2,
    "originalTitle": "Star Wars: Episode VII - The Force Awakens",
    "title": "Звёздные войны",
    "year": 2016,
  },
]
`;

exports[`kinopoisk/filmIdFromSearchResults filters results 4`] = `
Array [
  Object {
    "countries": Array [],
    "id": 2,
    "title": "foo",
    "year": 2016,
  },
  Object {
    "countries": Array [],
    "id": 4,
    "title": "foo",
    "year": 2016,
  },
]
`;

exports[`kinopoisk/filmIdFromSearchResults filters results 5`] = `
Array [
  Object {
    "countries": Array [
      "США",
      "Россия",
      "Германия",
    ],
    "id": 2,
    "title": "foo",
    "year": 2016,
  },
  Object {
    "countries": Array [
      "США",
      "Россия",
    ],
    "id": 4,
    "title": "foo",
    "year": 2016,
  },
]
`;
